docs/architecture/architecture.md

```diff
@@ -0,0 +1,671 @@
+  reaper into `cc-proxy` stream frames:
+```
+                  1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+0 1 2 3 4 5 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+┌───────────────────────────┬───────────────┬───────────────┐
+│          Version          │ Header Length │   Reserved    │
+├───────────────────────────┼─────┬─┬───────┼───────────────┤
+│          Reserved         │ Res.│E│  0x2  │    Opcode     │
+├───────────────────────────┴─────┴─┴───────┴───────────────┤
+│                      Payload Length                       │
+├───────────────────────────────────────────────────────────┤
+│                                                           │
+│                         Payload                           │
+│                                                           │
+│      (variable length, optional and opcode-specific)      │
+│                                                           │
+└───────────────────────────────────────────────────────────┘
+```
+- De-encapsulates and assembles standard output and error `cc-proxy` stream frames
+  into an output stream that it forwards to the container process reaper.
+- Translates all UNIX signals (except `SIGKILL` and `SIGSTOP`) into `cc-proxy`
+  `Signal` frames that it sends to the container via `cc-proxy`.
+
+As an example, assuming that running the `pwd` command from a containers standard
+input will generate the output "`/tmp`" on the containers standard output. If the
+`cc-agent` assigned to this process uses sequence number 8888 for stdin and stdout
+and 8889 for stderr, with `cc-shim` and Clear Containers, this example would look
+like:
+
+![cc-shim](arch-images/shim.png)
+
+## Networking
+
+Containers will typically live in their own, possibly shared, networking namespace.
+At some point in a container lifecycle, container engines will set up that namespace
+to add the container to a network which is isolated from the host network, but
+which is shared between containers
+
+In order to do so, container engines will usually add one end of a `virtual ethernet
+(veth)` pair into the container networking namespace. The other end of the `veth`
+pair
+is added to the container network.
+
+This is a very namespace-centric approach as QEMU can not handle `veth` interfaces.
+Instead it typically creates `TAP` interfaces for adding connectivity to a virtual
+machine.
+
+To overcome that incompatibility between typical container engines expectations
+and virtual machines, `cc-runtime` networking transparently bridges `veth`
+interfaces with `TAP` ones:
+
+![Clear Containers networking](arch-images/network.png)
+
+Clear Containers supports both
+[CNM](https://github.com/docker/libnetwork/blob/master/docs/design.md#the-container-network-model)
+and [CNI](https://github.com/containernetworking/cni) for networking management.
+
+### CNM
+
+![High-level CNM Diagram](arch-images/CNM_overall_diagram.png)
+
+__CNM lifecycle__
+
+1.  RequestPool
+
+2.  CreateNetwork
```

Comment:
These should all be rendered in a fixed-format using backticks.

Reply:
+1